File: hooks/lib/repo-root.js
/**
 * Purpose: Resolves the repo-root directory to anchor `.agentic/` state
 *          writes, instead of trusting the harness-supplied payload `cwd`
 *          verbatim. Prevents phantom `.agentic/` trees being written at
 *          whatever directory a stray `cd` (or a payload `cwd` that drifts
 *          across Bash tool calls) happens to leave the process in.
 *
 * Public API: resolveAgenticCwdWithDiagnostics(startDir) -> { root, driftLevels, foundGitAncestor }
 *             resolveAgenticCwd(startDir) -> string
 *
 * Upstream deps: node:fs (realpathSync, existsSync), node:path
 *
 * Downstream consumers: hooks/lib/wrap-marker.js, hooks/stop-context.js,
 *   hooks/pre-tool-use-spawn-emit.js, hooks/subagent-stop-spawn-emit.js,
 *   hooks/conductor-overreach-nudge.js, hooks/post-tool-use-capture-nudge.js,
 *   hooks/session-end-wrap.js, hooks/wrap-daemon.js,
 *   hooks/lib/skill-candidate-detector.js, hooks/lib/state-mark.js,
 *   hooks/lib/capture-gap.js, hooks/lib/context-rollup.js,
 *   bin/ds-wrap-acquire-lock, bin/ds-wrap-release-lock
 *
 * Failure modes: never throws. EACCES/ENOENT at any level along the walk is
 *   treated as "not found here, keep walking". If no `.git` ancestor is
 *   found within MAX_DEPTH, returns the realpath'd startDir unchanged with
 *   foundGitAncestor:false - callers must treat that as a resolution
 *   failure and SKIP the write, never silently write at the fallback path.
 *
 * Performance: a handful of fs.existsSync calls per invocation (at most
 *   MAX_DEPTH), no subprocess, no network.
 */

'use strict';

const fs = require('fs');
const path = require('path');

const MAX_DEPTH = 64;

/**
 * Realpath-pin startDir and walk up looking for a `.git` entry
 * (file or directory - EXISTENCE ONLY, never isDirectory()/isFile()).
 * A linked git worktree's `.git` is a FILE, not a directory, so a
 * dir-only check fails in the most common execution environment here.
 */
function resolveAgenticCwdWithDiagnostics(startDir) {
  let real;
  try {
    real = fs.realpathSync(startDir);
  } catch (_err) {
    real = startDir;
  }

  let current = real;
  let driftLevels = 0;

  for (let i = 0; i <= MAX_DEPTH; i += 1) {
    let hasGit = false;
    try {
      hasGit = fs.existsSync(path.join(current, '.git'));
    } catch (_err) {
      hasGit = false;
    }

    if (hasGit) {
      return { root: current, driftLevels, foundGitAncestor: true };
    }

    const parent = path.dirname(current);
    if (parent === current) {
      // Reached filesystem root without finding a `.git` ancestor.
      break;
    }
    current = parent;
    driftLevels += 1;
    if (driftLevels > MAX_DEPTH) {
      break;
    }
  }

  return { root: real, driftLevels: 0, foundGitAncestor: false };
}

function resolveAgenticCwd(startDir) {
  return resolveAgenticCwdWithDiagnostics(startDir).root;
}

module.exports = {
  resolveAgenticCwdWithDiagnostics,
  resolveAgenticCwd,
  MAX_DEPTH,
};

File: hooks/lib/repo_root.py
"""
Purpose: Resolves the repo-root directory to anchor `.agentic/` state
         writes, instead of trusting the harness-supplied payload `cwd`
         verbatim. Prevents phantom `.agentic/` trees being written at
         whatever directory a stray `cd` (or a drifted payload `cwd`)
         happens to leave the process in. Mirrors hooks/lib/repo-root.js.

Public API: resolve_agentic_cwd_with_diagnostics(start_dir) -> dict with
                "root", "drift_levels", "found_git_ancestor"
            resolve_agentic_cwd(start_dir) -> str

Upstream deps: os.path (realpath, exists, join, dirname)

Downstream consumers: hooks/lib/loop_guard.py, hooks/lib/enforcement_log.py,
    hooks/enforce-no-abdication.py, hooks/enforce-turn-shape.py,
    hooks/enforce-skeptic-round-cap.py, hooks/enforce-planning-artifact-spawn.py,
    bin/ds-status, bin/ds-cost, bin/ds-memory (all three via a lazy
    importlib.util dynamic loader, not a direct import)

Failure modes: never raises. Any OSError (EACCES/ENOENT) while probing a
    given level is treated as "not found here, keep walking". If no
    `.git` ancestor is found within MAX_DEPTH, returns the realpath'd
    start_dir unchanged with found_git_ancestor=False - callers must treat
    that as a resolution failure and SKIP the write, never silently write
    at the fallback path.

Performance: a handful of os.path.exists calls per invocation (at most
    MAX_DEPTH), no subprocess, no network.
"""

from __future__ import annotations

import os

MAX_DEPTH = 64


def resolve_agentic_cwd_with_diagnostics(start_dir: str) -> dict:
    """Realpath-pin start_dir and walk up looking for a `.git` entry
    (file or directory - EXISTENCE ONLY, never os.path.isdir()).
    A linked git worktree's `.git` is a FILE, not a directory, so a
    dir-only check fails in the most common execution environment here.
    """
    try:
        real = os.path.realpath(start_dir)
    except OSError:
        real = start_dir

    current = real
    drift_levels = 0

    for _ in range(MAX_DEPTH + 1):
        has_git = False
        try:
            has_git = os.path.exists(os.path.join(current, ".git"))
        except OSError:
            has_git = False

        if has_git:
            return {
                "root": current,
                "drift_levels": drift_levels,
                "found_git_ancestor": True,
            }

        parent = os.path.dirname(current)
        if parent == current:
            break
        current = parent
        drift_levels += 1
        if drift_levels > MAX_DEPTH:
            break

    return {"root": real, "drift_levels": 0, "found_git_ancestor": False}


def resolve_agentic_cwd(start_dir: str) -> str:
    return resolve_agentic_cwd_with_diagnostics(start_dir)["root"]
